### DISCUSSION TITLE
Mocking LucidORM

### DISCUSSION THREAD

**Discussion by nickabb** [adonisjs/core#4248] [Japa]
I'd like to be able to mock the AdonisJS models in order to not have to hit a live database for unit tests.  I understand your comments on [mocking](https://github.com/adonisjs/core/discussions/1612) - but I think we'll have to agree to disagree here - I'd still like to be able to do this in order to speed up tests.

Passing in each model as a parameter and then overriding them for tests works, but I'm trying to avoid that (this is the point of using an IoC container after all).

How can I do this?

**Comment by RomainLanz** [comment]
Hey @nickabb! 👋🏻 

That would be way easier to create an abstraction and change that abstraction during your test.

You can follow the repository pattern, for example.

```ts
class LucidUserRepository {
  getAll() { return User.all() }
}
```
```ts
const users = []
class InMemoryUserRepository {
  getAll() { return users }
}
```
```ts
abstract class UserRepository {
  abstract getAll()
}
```

Register inside a provider the abstraction you want to use:

```ts
export default class AppProvider {
  constructor(protected app: ApplicationContract) {}
  
  public async boot() {
    const LucidUserRepository = await import('infrastructure/repositories/user.lucid.repository')
    this.app.container.bind('UserRepository', () =>
      this.app.container.make(LucidUserRepository.default)
    )
  }
}
```

Then inject the class using the IoC Container:

```ts
@inject(['UserRepository'])
class UserController {
  constructor (private repository: UserRepository) {}
  
  index() {
    return this.repository.getAll()  
  }
}
```

Swap the implementation during testing:

```ts
test.group('User', (group) => {
  group.setup(() => {
    Application.container.useProxies(true)

    return () => {
      Application.container.useProxies(false)
    }
  })
  
  group.each.setup(async () => {
    return () => Application.container.restore()
  })
  
  test('...', async ({ assert }) => {
    Application.container.fake('UserRepository', () => {
      return new InMemoryUserRepository()
    }))
    
    // ...
  })
}
```

This flow will be greatly simplified in V6.

**Reply by nickabb** [reply]
Thanks for this!

First thoughts is that this is sort of what I was trying to avoid (a ton of boilerplate just for tests) - I'd have to write an abstraction over every model to call through to a library model  - and there would be a bunch of refactoring to inject the repositories into the constructors of each controller (not horrible, but still a tangible amount of work that compounds with project size)

You mentioned v6 - what will this look like in v6? Possible there are any docs so I could understand further if it would be worth an early upgrade?

**Reply by RomainLanz** [reply]
In AdonisJS 6, you will not have to use a string-based mapping.
You can watch [our live stream](https://www.youtube.com/watch?v=XXBwZiIQB98) on the subject to learn more about the new IoC.

Concerning your request, there is no other way to do it properly. Abstraction is needed to cleanly "mock" your dependency.
